### PR TITLE
Mask padded MiniLLM completion tokens out of the advantage and its length

### LINKED ISSUE #7024: MiniLLM: right-padding leaks into the reverse-KL advantage, and length normalization depends on batch padding
### Reproduction

Both points below are on `main` and are independent of #6635, which only changes how the discounted sum is computed. They came up while reviewing that PR.

**1. Right-padding is never masked out of the advantage.** `MiniLLMTrainer.compute_loss` builds the mask it hands to `_compute_advantage` from the raw completion ids:

```python
shifted_labels = input_ids[:, prompt_lengths:]
...
mask = shifted_labels != -100
```

`input_ids` never contains `-100`; the `-100` fill is applied to `labels`, a separate tensor built a few lines earlier from `attention_mask`. So `mask` is all ones, padded positions keep a reward of `teacher_logp - student_logp` on the pad token, and every earlier position's advantage sums those padded rewards in. With one 4-slot completion whose last two slots are padding, per-token rewards `teacher_logp - student_logp` of `[1.5, 1.0, 2.0, 2.0]` and `gamma=1.0`, the mask the caller builds and the mask it should build give different advantages at every real position:

```
caller mask       [1, 1, 1, 1]   advantages [6.5, 5.0, 4.0, 2.0]
completion mask   [1, 1, 0, 0]   advantages [2.5, 1.0, 0.0, 0.0]
```

The fix is a one-line change, `shifted_labels = labels[:, prompt_lengths:]`, since `labels` already carries the `-100` fill.

**2. Under `length_normalization=True` the result depends on how much padding the batch carries.** `_compute_advantage` replaces every masked position with `1e-4` before building the denominator, so the denominator of a short completion includes `1e-4 * gamma^(i-t)` for each pad slot after it. With a constant reward of `1.0` and `gamma=1.0`, which should give an advantage of exactly `1.0` everywhere, the value at the first position moves with the padded length:

```
padded to T=512   L=1     first=0.9514   last_valid=0.9514
                  L=128   first=0.9997   last_valid=0.9630
                  L=512   first=1.0000   last_valid=1.0000
```

The same completion gets a different advantage depending on the longest completion in its batch. The `1e-4` exists so that a fully masked row does not divide by zero; a `clamp(min=...)` on the finished denominator would give that guarantee without adding the pad slots to every real position's length.

Both tables come from calling the method directly with a stub `self`, the way `tests/experimental/test_minillm_trainer.py` does:

```python
import torch, types
from trl.experimental.minillm.minillm_trainer import MiniLLMTrainer

student = torch.tensor([[-2.0, -1.5, -3.0, -3.0]])
teacher = torch.tensor([[-0.5, -0.5, -1.0, -1.0]])
stub = types.SimpleNamespace(gamma=1.0, length_normalization=False)
print(MiniLLMTrainer._compute_advantage(stub, student, teacher, torch.ones(1, 4)))
print(MiniLLMTrainer._compute_advantage(stub, student, teacher, torch.tensor([[1.0, 1.0, 0.0, 0.0]])))

stub = types.SimpleNamespace(gamma=1.0, length_normalization=True)
for length in (1, 128, 512):
    mask = torch.zeros(1, 512); mask[0, :length] = 1
    adv = MiniLLMTrainer._compute_advantage(stub, torch.zeros(1, 512), torch.ones(1, 512), mask)[0]
    print(length, adv[0].item(), adv[length - 1].item())
```

### System Info

- Platform: Linux 5.14.0 (x86_64)
- Python 3.11
- TRL: 1.10.0 (dev, `main`)
- PyTorch 2.11.0
- Transformers 5.11.0
- Accelerate 1.13.0

### Checklist

- [x] I have checked that my issue isn't already filed
- [x] I have included my system information
- [x] Any code provided is minimal, complete, and reproducible
- [x] Any code provided is properly formatted in code blocks
- [x] Any traceback provided is complete

### PR DESCRIPTION
**Pull request by behroozazarkhalili**
MiniLLM included padded tokens in advantages and let padding change the length normalization denominator. I built the advantage mask from labels and counted only unmasked positions in the discounted length, with a guard for empty positions. Both new regression tests fail on main, and the full MiniLLM test file passes on this branch.
Fixes #7024.

---

> [!NOTE]
> **Medium Risk**
> Changes experimental MiniLLM training math (advantage masking and normalization), which directly affects loss gradients but is a targeted correctness fix with new regression tests.
> 
> **Overview**
> Fixes incorrect reverse-KL advantages in **MiniLLM** when batches use right-padded completions.
> 
> **`compute_loss`** now builds the completion mask from **`labels`** (where padding is `-100`) instead of **`shifted_labels`** from **`input_ids`**, so padded completion slots no longer contribute rewards or bleed into earlier positions’ advantages. Log-prob gathers still use **`input_ids`** indices.
> 
> **`_compute_advantage`** length normalization no longer treats masked positions as `1e-4` in the discounted-length denominator. Only unmasked tokens count, so short completions padded to a fixed width get the correct normalized advantage (e.g. 1.0 for constant reward with `gamma=1.0`). Zero-length tails divide by 1 instead of using a floor that would distort late valid tokens when `gamma < 1`.
> 
> Adds hermetic tests for padding-invariant length normalization and for the mask passed into **`_compute_advantage`** from **`compute_loss`**.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit 9ccdf8f7ad1ee440279a8a16dcc833d2520074da. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

File: trl/experimental/minillm/minillm_trainer.py [hunk 1]
```diff
@@ -336,10 +336,10 @@ def _compute_advantage(
...
 
             if self.length_normalization:
-                mask = torch.where(mask < 0.5, 1e-4, mask)
-                lengths = mask * gamma_pow
-                lengths = lengths.flip(1).cumsum(dim=1).flip(1)
-                advantages = advantages / lengths
+                # Only unmasked positions count toward the discounted length, so the advantage of a completion does
+                # not depend on how much padding the batch carries. The clamp keeps a fully masked row finite.
+                lengths = (mask * gamma_pow).flip(1).cumsum(dim=1).flip(1)
+                advantages = advantages / lengths.clamp(min=1e-4)
```
[the remark this answers is not in the document]
Comment (behroozazarkhalili) [reply trl/experimental/minillm/minillm_trainer.py]: Confirmed by execution: with gamma 0.9, a constant reward of 1.0 and an unpadded 512-token row, position 300 got 1.9e-9 and position 511 got 4.1e-20 instead of 1.0, because the floor of 1e-4 exceeds the discounted length of late positions. Fixed in 7a573e60: the divisor is now `lengths.masked_fill(lengths == 0, 1.0)`. A length is zero only where every later position is masked, and there the numerator is zero as well, so the guard no longer touches valid positions. `test_length_normalization_ignores_padding` now runs at gamma 1.0 and 0.9 and asserts the advantage equals the reward at every valid position; the floored version fails it at 19 of 128 positions.

**Comment by behroozazarkhalili** [comment]
Closing this one myself. It came from reading the code, not from a failure in a real training run, and per the discussion on #6862 that isn't the bar for this repo. If someone hits this in practice, the branch is still here and I'm happy to reopen it with a reproduction.
